io/jvm/src/main/scala/scalaz/effect/RTS.scala

```diff
@@ -278,9 +273,13 @@ private object RTS {
       // was caught but the stack is empty, we make the stack non-empty. This
       // lets us return only the finalizer, which will be null for common cases,
       // and result in zero heap allocations for the happy path.
-      if (caught && stack.isEmpty) stack.push(IdentityCont)
+      if (caught) {
```

Comment:
`caught` is now redundant with `!(errorHandler eq null)`, so you don't need both. I'd delete `caught`.